### PR TITLE
ISPN-6073 Replace Set<Flag> by a bit set

### PR DESCRIPTION
**Pull request by pruivo**
https://issues.jboss.org/browse/ISPN-6073

API change: LocalFlagAffectedCommand

### CODE REVIEW

**Comment by tristantarrant** [comment]
Nice. How much does this change in terms of allocation costs ?

**Comment by pruivo** [comment]
@tristantarrant you have the results in the mailing list with the flight recorder snapshots. 
The resume is (1 min run):
```
Allocations (Master):
Long:           1356 allocations, 187.51 MB
RegularEnumSet: 1762 allocations, 126.45 MB

Allocations (BitSet)
Long:           1677 (+321) allocations, 211.70 (+24.19) MB
RegularEnumSet: 0 (-1762), 0 (-126.45) MB
```

**Comment by danberindei** [comment]
@pruivo there's something off with those numbers, there's no way 321 Longs can use 24MB...

**Comment by pruivo** [comment]
@danberindei from my little understanding in Flight Recorder, it is the TLAB allocations, not the object allocations.

**Comment by danberindei** [comment]
@pruivo Ok, looks like they aren't counting all allocations, they are only sampling the allocations that don't fit in the current [TLAB][1]. I suppose they still count everything when it comes to the total size, although I'm not sure how they do it...

Anyway, the number and size of allocations per minute will depend not just on the benchmark you're running, but also on the machine you're running it on. JMH has a `gc` profiler that you can enable with `-prof gc` and that gives you the allocation rate per operation. I haven't used it enough to confirm that it's reliable for multithreaded, multi-node tests, but it's on my list of things to try :)

[1]: http://hirt.se/blog/?p=381

File: core/src/main/java/org/infinispan/commands/AbstractLocalFlagAffectedCommand.java [hunk 1]
```diff
@@ -12,15 +11,23 @@
  */
 public abstract class AbstractLocalFlagAffectedCommand implements LocalFlagAffectedCommand {
 
-   protected Set<Flag> flags;
+   private long flags = 0;
 
    @Override
-   public Set<Flag> getFlags() {
+   public long getFlagsBitSet() {
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/commands/AbstractLocalFlagAffectedCommand.java]: Couldn't you keep the old method and allocate an `EnumSet` on demand?

File: core/src/main/java/org/infinispan/commands/CommandsFactory.java [hunk 2]
```diff
@@ -91,39 +90,41 @@
     * @param key key to put
     * @param value value to put
     * @param metadata metadata of entry
-    * @param flags Command flags provided by cache
+    * @param flagsBitSet Command flags provided by cache
     * @return a PutKeyValueCommand
     */
-   PutKeyValueCommand buildPutKeyValueCommand(Object key, Object value, Metadata metadata, Set<Flag> flags);
+   PutKeyValueCommand buildPutKeyValueCommand(Object key, Object value, Metadata metadata, long flagsBitSet);
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/commands/CommandsFactory.java]: Based on the package, I think `CommandsFactory` is public API too.

File: core/src/main/java/org/infinispan/commands/functional/ReadWriteManyCommand.java [hunk 3]
```diff
@@ -180,11 +181,21 @@ public boolean alwaysReadsExistingValues() {
    }
 
    @Override
+   public long getFlagsBitSet() {
+      return EnumUtil.EMPTY_BIT_SET;
+   }
+
+   @Override
    public void setFlags(Set<Flag> flags) {
       // TODO: Customise this generated block
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/commands/functional/ReadWriteManyCommand.java]: Remove TODO?

File: core/src/main/java/org/infinispan/commands/LocalFlagAffectedCommand.java [hunk 4]
```diff
@@ -17,14 +19,22 @@
     * not be modified directly, only via the {@link #setFlags(Set)}, {@link #addFlag(Flag)} and {@link
     * #addFlags(Set)} methods.
     */
-   Set<Flag> getFlags();
+   default Set<Flag> getFlags() {
+      return EnumUtil.enumSetOf(getFlagsBitSet(), Flag.class);
+   }
+
+   long getFlagsBitSet();
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/commands/LocalFlagAffectedCommand.java]: If this is supposed to be backwards-compatible, then you need to have a default implementation for `getFlagsBitSet()` based on `getFlags()` now, and only switch them in 9.0.

File: core/src/main/java/org/infinispan/commands/CommandsFactory.java (hunk 2 above)
Comment (pruivo) [inline core/src/main/java/org/infinispan/commands/CommandsFactory.java]: that's why I marked as a Preview a not for 8.2 :)

File: core/src/main/java/org/infinispan/commands/functional/ReadWriteManyCommand.java (hunk 3 above)
Comment (pruivo) [inline core/src/main/java/org/infinispan/commands/functional/ReadWriteManyCommand.java]: No idea why the TODOs are there, but there are a bunch of methods like this. Not sure if the class is completed or not.

File: core/src/main/java/org/infinispan/commands/AbstractLocalFlagAffectedCommand.java (hunk 1 above)
Comment (pruivo) [inline core/src/main/java/org/infinispan/commands/AbstractLocalFlagAffectedCommand.java]: the getFlags is not removed. It is implemented in the interface.
Yes, it was the only way I had to copy the flags to another commands (e.g. cluster get to lock control)

File: core/src/main/java/org/infinispan/commands/read/GetCacheEntryCommand.java [hunk 5]
```diff
@@ -25,9 +23,8 @@
    private InternalEntryFactory entryFactory;
    private InternalCacheEntry remotelyFetchedValue;
 
-   public GetCacheEntryCommand(Object key, Set<Flag> flags, InternalEntryFactory entryFactory) {
-      this.key = key;
-      this.flags = flags;
+   public GetCacheEntryCommand(Object key, long flagsBitSet, InternalEntryFactory entryFactory) {
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/commands/read/GetCacheEntryCommand.java]: Do we need to keep the old constructor for backwards compatibility?

File: core/src/main/java/org/infinispan/context/Flag.java [hunk 6]
```diff
@@ -272,64 +270,10 @@
...
 
    /**
-    * Creates a copy of a Flag Set removing instances of FAIL_SILENTLY.
-    * The copy might be the same instance if no change is required,
-    * and should be considered immutable.
-    * @param flags
-    * @return might return the same instance
+    * Creates a copy of a Flag BitSet removing instances of FAIL_SILENTLY.
     */
-   public static Set<Flag> copyWithoutRemotableFlags(Set<Flag> flags) {
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/context/Flag.java]: We should just mark the old methods as deprecated.

File: core/src/main/java/org/infinispan/commands/functional/ReadWriteManyCommand.java (hunk 3 above)
Comment (slaskawi) [inline core/src/main/java/org/infinispan/commands/functional/ReadWriteManyCommand.java]: +1 for killing todo :) Comments die last :)

**Comment by tristantarrant** [comment]
Pushed, thanks

**Pull request merged**